### ISSUE TITLE
Optimizer rule reduce-extraction-to-projection breaks queries

### ISSUE THREAD

**Issue by Simran-B** [arangodb/arangodb#3011]
## my environment running ArangoDB
- ArangoDB v3.2.0
- Mode: Single-Server
- Storage-Engine: rocksdb
-  OS: Windows 10 Professional 64bit

I'm issuing AQL via:
- web interface with this browser: Chrome 59.0.3071.115 

Explain:
```
Execution plan:
 Id   NodeType                  Est.   Comment
  1   SingletonNode                1   * ROOT
  2   EnumerateCollectionNode     10     - FOR doc IN products   /* full collection scan, projection: `salesrank` */
  3   CalculationNode             10       - LET #2 = MERGE(doc, { "salesrank" : TO_NUMBER(doc) })   /* simple expression */   /* collections used: doc : products */
  4   InsertNode                   0       - INSERT #2 IN copy

Indexes used:
 none

Optimization rules applied:
 Id   RuleName
  1   remove-data-modification-out-variables
  2   reduce-extraction-to-projection
```

The AQL query in question is:
```js
FOR doc IN products
    INSERT MERGE(
        doc,
        {salesrank: TO_NUMBER(doc.`salesrank`)}
    ) INTO copy
```

The issue can be reproduced using this dataset: [products.zip](https://github.com/arangodb/arangodb/files/1211584/products.zip) (can be imported via Web UI)

- The following problem occurs:
  > Query: invalid document type (while executing)

- Instead I would be expecting:
  Successful execution and all products copied over to collection `copy`, with attribute `salesrank` casted from string to number.

The execution plan doesn't look correct. It passes `doc` into `TO_NUMBER()`, although it should pass the `salesrank` attribute.  The optimizer rule `reduce-extraction-to-projection` is apparently too eager, the query works if it is disabled:

```js
db._createStatement({
"query": "FOR doc IN products INSERT MERGE(doc, {salesrank: TO_NUMBER(doc.salesrank)}) INTO copy",
"options": {"optimizer": {"rules": ["-reduce-extraction-to-projection"]}}
}).execute()
```

Also affected: `FOR doc IN products RETURN MERGE(doc, {salesrank: TO_NUMBER(doc.salesrank)})`

Not affected, because the rule is not applied if more than a single attribute is taken from the document:
`FOR doc IN products RETURN MERGE(doc, {salesrank: TO_NUMBER(doc.salesrank), ASIN: doc.ASIN})`

**Comment by jsteemann** [comment]
Fixed in 3.2.1

**Comment by jsteemann** [comment]
3.2.1 has been released and is [available for download](https://www.arangodb.com/download-major/).
